File: plugins/trace/buffer.go
package trace

import (
	"sync"
)

const defaultBufferSize = 10000

// RingBuffer 是一个线程安全的环形缓冲区
// 用于暂存 Span 数据，支持溢出时丢弃最旧数据
type RingBuffer struct {
	mu           sync.Mutex
	data         []*Span
	head         int   // 读取位置
	tail         int   // 写入位置
	count        int   // 当前元素数量
	capacity     int   // 容量
	droppedCount int64 // 丢弃计数
}

// NewRingBuffer 创建指定大小的 RingBuffer
func NewRingBuffer(size int) *RingBuffer {
	if size <= 0 {
		size = defaultBufferSize
	}
	return &RingBuffer{
		data:     make([]*Span, size),
		capacity: size,
	}
}

// Push 将 span 添加到缓冲区
// 如果缓冲区已满，会丢弃最旧的数据
// 返回 false 如果 span 为 nil
func (b *RingBuffer) Push(span *Span) bool {
	if span == nil {
		return false
	}

	b.mu.Lock()
	defer b.mu.Unlock()

	// 如果缓冲区已满，丢弃最旧的数据
	if b.count == b.capacity {
		b.head = (b.head + 1) % b.capacity
		b.count--
		b.droppedCount++
	}

	// 写入数据
	b.data[b.tail] = span
	b.tail = (b.tail + 1) % b.capacity
	b.count++

	return true
}

// Flush 从缓冲区取出最多 n 个 span
// 返回的 span 按 FIFO 顺序排列
func (b *RingBuffer) Flush(n int) []*Span {
	b.mu.Lock()
	defer b.mu.Unlock()

	if b.count == 0 || n <= 0 {
		return nil
	}

	// 确定要取出的数量
	toFlush := n
	if toFlush > b.count {
		toFlush = b.count
	}

	result := make([]*Span, toFlush)
	for i := 0; i < toFlush; i++ {
		result[i] = b.data[b.head]
		b.data[b.head] = nil // 帮助 GC
		b.head = (b.head + 1) % b.capacity
		b.count--
	}

	return result
}

// Len 返回当前缓冲区中的元素数量
func (b *RingBuffer) Len() int {
	b.mu.Lock()
	defer b.mu.Unlock()
	return b.count
}

// Cap 返回缓冲区容量
func (b *RingBuffer) Cap() int {
	return b.capacity
}

// DroppedCount 返回因缓冲区满而丢弃的 span 数量
func (b *RingBuffer) DroppedCount() int64 {
	b.mu.Lock()
	defer b.mu.Unlock()
	return b.droppedCount
}

// Clear 清空缓冲区
func (b *RingBuffer) Clear() {
	b.mu.Lock()
	defer b.mu.Unlock()

	for i := range b.data {
		b.data[i] = nil
	}
	b.head = 0
	b.tail = 0
	b.count = 0
}

File: pocketless/src/plugins/trace/buffer.ts
/**
 * Trace 插件 — RingBuffer
 * 对照 Go 版 plugins/trace/buffer.go
 *
 * 线程安全的环形缓冲区，用于暂存 Span 数据。
 * 溢出时丢弃最旧数据（FIFO），并记录丢弃计数。
 */

import type { Span } from "./register";

const DEFAULT_BUFFER_SIZE = 10000;

/**
 * RingBuffer — 固定容量的环形缓冲区。
 *
 * 写满时自动覆盖最旧条目，并累计 droppedCount。
 * 通过 flush(n) 按 FIFO 顺序取出最多 n 个 Span。
 */
export class RingBuffer {
  private data: (Span | null)[];
  private head = 0;     // 读取位置
  private tail = 0;     // 写入位置
  private count = 0;    // 当前元素数量
  private readonly capacity: number;
  private dropped = 0;  // 因缓冲区满而丢弃的数量

  constructor(size: number = DEFAULT_BUFFER_SIZE) {
    this.capacity = size > 0 ? size : DEFAULT_BUFFER_SIZE;
    this.data = new Array(this.capacity).fill(null);
  }

  /** 写入 span，若满则丢弃最旧条目 */
  push(span: Span): boolean {
    if (!span) return false;

    // 缓冲区满时，覆盖最旧数据
    if (this.count === this.capacity) {
      this.head = (this.head + 1) % this.capacity;
      this.count--;
      this.dropped++;
    }

    this.data[this.tail] = span;
    this.tail = (this.tail + 1) % this.capacity;
    this.count++;
    return true;
  }

  /** 取出最多 n 个 span（FIFO 顺序），取出后从缓冲区移除 */
  flush(n: number): Span[] {
    if (this.count === 0 || n <= 0) return [];

    const toFlush = Math.min(n, this.count);
    const result: Span[] = [];

    for (let i = 0; i < toFlush; i++) {
      result.push(this.data[this.head] as Span);
      this.data[this.head] = null;
      this.head = (this.head + 1) % this.capacity;
      this.count--;
    }

    return result;
  }

  /** 清空所有条目 */
  clear(): void {
    for (let i = 0; i < this.capacity; i++) {
      this.data[i] = null;
    }
    this.head = 0;
    this.tail = 0;
    this.count = 0;
  }

  len(): number {
    return this.count;
  }

  cap(): number {
    return this.capacity;
  }

  /** 因缓冲区满而丢弃的 span 总数 */
  droppedCount(): number {
    return this.dropped;
  }
}
